core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/ValueStore.java

```diff
@@ -191,12 +200,62 @@ public NativeValue getValue(int id) throws IOException {
 			if (data != null) {
 				resultValue = data2value(id, data);
 
-				// Store value in cache
-				valueCache.put(cacheID, resultValue);
+				if (!(resultValue instanceof CorruptValue)) {
+					// Store value in cache
+					valueCache.put(cacheID, resultValue);
+				}
 			}
 		}
 
 		return resultValue;
+
+	}
+
+	/**
+	 * Gets the Resource for the specified ID.
+	 *
+	 * @param id A value ID.
+	 * @return The Resource for the ID, or <var>null</var> no such value could be found.
+	 * @throws IOException If an I/O error occurred.
+	 */
+	public <T extends NativeValue & Resource> T getResource(int id) throws IOException {
+
+		NativeValue resultValue = getValue(id);
+
+		if (resultValue != null && !(resultValue instanceof Resource)) {
+			if (SOFT_FAIL_ON_CORRUPT_DATA_AND_REPAIR_INDEXES && resultValue instanceof CorruptValue) {
+				return (T) new CorruptIRIOrBNode(revision, id, ((CorruptValue) resultValue).getData());
+			}
+			logger.warn(
+					"Possible corrupt data consider setting the system property org.eclipse.rdf4j.sail.nativerdf.softFailOnCorruptDataAndRepairIndexes to true");
```

Comment:
this message sounds a bit strange (though it is clear what it means. Maybe you can phrase it in two sentences, same in line 255 and line 629